draft-ietf-httpbis-priority.md

```diff
@@ -509,8 +509,54 @@ not specify the incremental(`i`) parameter.
 
 # Client Scheduling
 
-A client MAY use priority values to make local scheduling choices about the
-requests it initiates.
+A client MAY use priority values to make local processing or scheduling choices
+about the requests it initiates.
+
+
+# Server Scheduling
+
+Priority signals are input to a prioritization process. They do not guarantee
+any particular processing or transmission order for one response relative to any
+other response. An endpoint cannot force a peer to process concurrent request in
+a particular order using priority. Expressing priority is therefore only a
+suggestion.
+
+A server can use Priority signals along with other inputs to make scheduling
+decisions. No guidance is provided about how this can or should be done. Factors
+such as implementation choices or deployment environment also play a role. Any
+given connection is likely to have many dynamic permutations. For these reasons,
+there is no unilateral perfect scheduler and this document only provides some
+basic recommendations for implementations. 
+
+Clients can expect servers will make prioritization decisions, including
+ignoring all signals. And they should expect that decisions might be based on
+metadata or information beyond the scope of extensible priorities. 
+
+It is RECOMMENDED that, when possible, servers respect urgency values. Sending
+higher urgency responses before lower urgency responses.
+
+It is RECOMMENDED that, when possible, servers respect incremental values.
+Sending non-incremental responses of the same urgency one-by-one, and sending
+incremental responses of the same urgency in round-robin manner.
+
+It is RECOMMENDED that a server considers request generation order as an input
+to prioritization. This can be determined from the Stream ID. A server that
+receives concurrent requests at the same urgency level might serve the responses
+one-by-one but it needs to pick an order. Serving the lowest Stream ID in a
+given urgency level can align well with clients usage of HTTP; such as user
+agents that load document trees where ordering is important.
+
+For non-incremental resources the total download time (time to first byte - time
+to last byte) is important. For incremental resources chunk download times are
+important, especially the first. A server that receives a mix of incremental and
+non-incremental at the same urgency is challenged to schedule two competing
+factors. An unbalanced scheduler might prefer one type over another, leading to
+sub-optimal loading and in the worst case starvation of one type. Servers are
+RECOMMENDED to avoid starvation but no specific method of doing so is prescribed.
```

Comment:
I don't get this point about unbalanced scheduling.  It seems to imply that there might be some other reason to balance between atomic and incremental resources, but I don't think that is the intent.

Reply:
From https://github.com/httpwg/http-extensions/issues/1232#issuecomment-663724137

>  Let's imagine two cases:
 > 1) At the same urgency level, a huge non-incremental file download has started, then a small incremental resource is requested.
>  2) At the same urgency level, an incremental hanging GET is waiting for response, while a non-incremental file download is requested.

An unbalanced scheduler might be designed to completely flush one type of resource before moving on the the other. Especially likely if resources are large in comparison to the BDP. This could starve the other type from ever getting a share. The text is attempting to say don't do that. To avoid it, an implemention could somehow yield sending one type in a given time period. Guoye's suggestion on https://github.com/httpwg/http-extensions/issues/1232#issuecomment-663889717 provides an example. I like to avoid recommending any specific solution.

Reply:
I don't think that you want to use the word unbalanced because it carries negative connotations.

What those examples are suggesting is a few questions:

Do you want to permit small responses to jump ahead of in-progress responses?
What if something that is nominally higher priority (by order, not urgency) can't start until a lower priority response has begun? Can it pre-empt?  Can it pre-empt only based on size?
(There's another one I thought of: Do you permit non-incremental responses to start when incremental responses are in progress? Maybe you have an answer for that already.)

These imply judgment being exercised or the presence of "other inputs".  But it is probably sensible to allow it, but that is clearly not the intent conveyed by the signal.  Maybe size-based discrimination is appropriate, but it's not something that this scheme supports, so you need to be careful.

What I would do then is to enumerate corner cases where we know that strict adherence to the scheme could end up with suboptimal results.  These cases are exactly the ones that I'd include there.

This scheme can't address these cases.  So you need a clear delineation between what following the scheme gets you and where you are using "special sauce".  This paragraph went from describing how this scheme operates and the consequences of that straight to special sauce stuff with no pause for breath.

Reply:
I've reworked this paragraph to address the points here and in the other comment. See https://github.com/httpwg/http-extensions/pull/1266/commits/a215d40b52b0d1ed5ae67235ee23aad49d75adf5

(my force push seems to have broken Github's tracking, I aplogize).